### PR TITLE
Expanding typing for issues.all function

### LINKED ISSUE #3384: Improve/fix `IssueSchema.labels` / `MergeRequestSchema.labels` typing
**Description**

`IssueSchema.labels` is typed as `undefined | string[]` which is correct unless `requestOptions.withLabelsDetails` is set to true.

In that case, Gitlab API does not return an array of string for this property, but an array of objects:

```typescript
const issues = await api.issues.all({ projectId: '...', withLabelsDetails: true });
```

```json5
[
  {
    id: 123,
    iid: 12345,
    // ...
    labels: [
      {
        id: 456,
        name: "bug",
        description: null,
        description_html: "",
        text_color: "#FFFFFF",
        color: "#000000",
      },
    ],
    // ...
  },
]
```

Merge requests have the same issue with the `withLabelDetails` request option.

**Proposal**

##### Easy fix: union type

Type `IssueSchema.labels` as an union : `string[] | IssueLabelDetailsSchema[]`.

This would generate typescript errors for current users.

```typescript
const issues = await api.Issues.all({ projectId: 'foo' });
for(const issue of issues) {
  for(const label of labels) {
    if(label.startsWith('foo')) { // Error !
      continue;
    }
    // ...
  }
}
```

##### Harder fix

IMO the ideal fix would be as follow:

```typescript
issues = await api.Issues.all({ projectId: 'foo' });
issues[0].labels; // undefined | string[]

issues = await api.Issues.all({ projectId: 'foo', withLabelDetails: true });
issues[0].labels; // undefined | IssueLabelDetailsSchema[];
```

I think I've got a solution with generics working for this kind of stuff (i.e, tweaking returned entity types based on request options). I'll follow this up with a draft PR for discussion.

**Checklist**
- [x] I have checked that this is not a duplicate issue.
- [x] I have read the documentation.

### PR DESCRIPTION
**Pull request by jdalrymple**
fixes #3384

### CODE REVIEW

File: packages/core/src/resources/Issues.ts [hunk 1]
```diff
@@ -24,6 +24,15 @@ export interface TimeStatsSchema extends Record<string, unknown> {
...
 
+export interface IssueLabelDetailsSchema extends Record<string, unknown> {
+  id: number;
+  name: string;
+  description: null | string;
+  description_html: string;
+  text_color: string;
+  color: string;
+}
+
```
Comment (kouak) [inline packages/core/src/resources/Issues.ts:35]: Issues, MergeRequest and Epics can all have LabelDetails attached to them.

This interface could probably be moved to `templates/ResourceLabel.ts` and renamed `LabelDetailsSchema`.

**Comment by kouak** [comment]
:+1:

This implementation appears to be working fine:

```typescript
let stringLabel: undefined | string;
let detailedLabel: undefined | IssueLabelDetailsSchema;

stringLabel = (await service.all({ groupId: 2 })).at(0)?.labels?.at(0);
// @ts-expect-error -- By default, labels are string
detailedLabel = (await service.all({ groupId: 2 })).at(0)?.labels?.at(0);

stringLabel = (await service.all({ groupId: 2, withLabelsDetails: false }))
  .at(0)
  ?.labels?.at(0);
// @ts-expect-error -- By default, labels are string
detailedLabel = (await service.all({ groupId: 2, withLabelsDetails: false }))
  .at(0)
  ?.labels?.at(0);

// @ts-expect-error -- In that case, labels are detailed labels
stringLabel = (await service.all({ groupId: 2, withLabelsDetails: true }))
  .at(0)
  ?.labels?.at(0);
detailedLabel = (await service.all({ groupId: 2, withLabelsDetails: true }))
  .at(0)
  ?.labels?.at(0);
```

FIY, I've quickly glanced over the codebase and gitlab API documentation, and I've found a few similar options, where the response :
- `Issues.withLabelsDetails`
- `MergeRequests.withLabelDetails`
- `Epics.withLabelDetails`
- `Users.withCustomAttributes`
- `Project.withCustomAttributes`
- `ProjectLabels.withCount`
- `GroupLabels.withCount`
- `GroupWiki.withContent`
- `Groups.withCustomAttributes`
- `Commits.withStats`
- Possibly `Commits.trailers` ?
- `Release.includeHtmlDescription`

Overloads should be fine for most (if not all) of those.

**Comment by jdalrymple** [comment]
Ill give it a look and include it! Thanks for the review

**Comment by jdalrymple** [comment]
Any chance you know what the payload for these?

1. Projects/Users/Groups.withCustomAttributes 
2. Commits.withStats
3. Commits.trailers

**Comment by kouak** [comment]
I've found an openapi definition file in gitlab repo : https://gitlab.com/gitlab-org/gitlab/-/blob/master/doc/api/openapi/openapi.yaml

- [Commits.trailer](https://gitlab.com/gitlab-org/gitlab/-/blob/master/doc/api/openapi/openapi.yaml?plain=1#L3059-3062): `Record<string, string>` ? Or maybe `Record<string, unknown>` ?
- [CustomAttributes](https://gitlab.com/gitlab-org/gitlab/-/blob/master/doc/api/openapi/openapi.yaml?plain=1#L2976-2984): `Array<{ key: string, value: string }>`

**Comment by kouak** [comment]
Could not find `Commit.stats` anywhere, but it appears to be of type `{ additions: number; deletions: number; total: number; }` based on my testing.

**Comment by kouak** [comment]
Yeah that's the right type for `Commits.stats`, here's the Gitlab source code responsible this : https://gitlab.com/gitlab-org/gitlab-foss/-/blob/master/lib/api/entities/commit_stats.rb

**Comment by jdalrymple** [comment]
I think i got them all. Give it a once over if you have a moment!

File: packages/core/src/resources/Commits.ts [hunk 2]
```diff
@@ -195,6 +197,27 @@ export type EditPipelineStatusOptions = {
...
+      PaginationRequestOptions<P> &
+      Sudo &
+      ShowExpanded<E> & { trailers: true },
+  ): Promise<GitlabAPIResponse<(CommitSchema & { trailers: Record<string, unknown> })[], C, E, P>>;
+
+  all<E extends boolean = false, P extends PaginationTypes = 'offset'>(
+    projectId: string | number,
+    options?: AllCommitsOptions & PaginationRequestOptions<P> & Sudo & ShowExpanded<E>,
+  ): Promise<GitlabAPIResponse<CommitSchema[], C, E, P>>;
+
```
Comment (kouak) [inline packages/core/src/resources/Commits.ts:220]: There should be another branch to cover the `{ trailers: true, withStats: true }` case.

image

**Review by kouak — COMMENTED** [review]
LGTM, only misses the `Commit.all(..., { trailers: true, withStats: true })`

**Comment by jdalrymple** [comment]
:rocket: PR was released in [`39.12.0`](https://github.com/jdalrymple/gitbeaker/releases/tag/39.12.0) :rocket:

**Pull request merged**